utils/src/invert_range.rs

```diff
@@ -0,0 +1,166 @@
+use std::ops::Range;
+use thiserror::Error;
+
+/// Inverts a range depending on a set of ranges, i.e. returns the complement
+///
+/// This function takes a range and a set of ranges to remove from the original range. For example
+/// if you provide the `original_range` `1..10` and the `ranges_to_remove` `2..5` and `7..8` the
+/// function will return the ranges `1..2`, `5..7` and `8..10`.
+///
+/// Every range in `ranges_to_remove` must be contained in `original_range`.
+/// Ranges must not overlap and cannot be empty or negative.
+pub fn invert_range<T: Ord + Copy>(
+    original_range: &Range<T>,
+    ranges_to_remove: &[Range<T>],
+) -> Result<Vec<Range<T>>, RangeError> {
+    // Check that original_range is valid
+    if original_range.start >= original_range.end {
+        return Err(RangeError::Invalid);
+    }
+
+    for (k, range) in ranges_to_remove.iter().enumerate() {
+        // Check that there is no invalid or empty range
+        if range.start >= range.end {
+            return Err(RangeError::Invalid);
+        }
+
+        // Check that ranges are not out of bounds
+        if range.start >= original_range.end
+            || range.end > original_range.end
+            || range.start < original_range.start
+            || range.end <= original_range.start
+        {
+            return Err(RangeError::OutOfBounds);
+        }
+
+        // Check that ranges are not overlapping
+        if ranges_to_remove
+            .iter()
+            .enumerate()
+            .any(|(l, r)| k != l && r.start < range.end && r.end > range.start)
+        {
+            return Err(RangeError::Overlapping);
+        }
+    }
+
+    // Now invert ranges
+    let mut inverted = vec![original_range.clone()];
+
+    for range in ranges_to_remove.iter() {
+        let inv = inverted
+            .iter_mut()
+            .find(|inv| range.start >= inv.start && range.end <= inv.end)
+            .expect("Should have found range to invert");
+
+        let original_end = inv.end;
+        inv.end = range.start;
+
+        inverted.push(Range {
+            start: range.end,
+            end: original_end,
+        });
+    }
```

Comment:
With `ranges_to_remove` sorted you can just operate on the tail

```suggestion
    let mut inverted = Vec::with_capacity(ranges_to_remove.len() + 1);
    inverted.push(original_range.clone());
    for remove in ranges_to_remove {
        let tail = inverted.last_mut().unwrap();
        
        tail.end = remove.start;
        
        inverted.push(remove.end..original_range.end);
    }
```

Reply:
very nice! :+1: